EIP-009-js-pdk-interface-definition.md

```diff
@@ -0,0 +1,135 @@
+# JS-PDK Interface Defintion
+
+## Purpose
+
+We need to support host functions in the JS PDK but the current way we define
+the interface, while working for exports, will not work for imports. Here is a rough
+outline of how the current system works:
+
+First we compile quickjs to wasm, we call this the "engine" or "core" module.
+This module is a rust wasm app with quickjs embedded and has one custom function as an export `__invoke`.
+`__invoke` takes an index (n), it looks at all the javascript exports in your main module, sorts them alphabetically,
+then invokes the (nth) one.
+
+The next step is to wizen and mutate this core module. The user passes in some javascript
+which we eval and wizen. We need need to expose the exports in the export section. We do this
+by evaling the js (in this compile step), reading the js module's exports, sorting them alphabetically,
+and adding them to the exports table in order. The code for these functions (we call thunks) are also generated 
+and just call `__invoke` with the export function id (where is sits in the order).
+
+There are some downsides to doing this for exports, it's complicated and slow, but it also works fine.
+This will not work for imports because we'd need to come up with a very wacky way to programatically extract
+and type the imports from the user's js code. We can no longer magically assume we will know the wasm signature.
+Exports are easy because all Extism exports have the same signature. Imports do not. Furthermore this process
+of mutating a wasm module and adding some exports and functions works because it is a fairly additive process.
+Importing on the other hand requires re-aligning a lot of other tables. And if i were to continue custom coding
+that I'd be building a linker.
+
+## Solution
+
+What we need to start doing is have the programmer define for us their interface explicitly. They'll be able to
+define the Exports and Imports in an IDL. We also need to use tooling to link the 
+core module and the user code module so we don't fall down the trap of havign to build a linker.
+
+Here is how I propose the new path will work. This may not be permanent but I think 
+it will work for 1.0:
+
+![js pdk pipeline](content/009-js-pdk-pipeline.png)
+
+The programmer will define their interface in a typescript file `interface.d.ts`:
+
+```typescript
+// all extism exports have type () -> i32
+interface Exports {
+  myExport(): I32;
+}
+
+// this interface is optional and types are  defined by you
+// to match your host functions
+interface Imports {
+  myHostFunc1(p: I64, q: I32): I64;
+  myHostFunc2(p: I64): I64;
+}
+```
```

Comment:
```suggestion
```typescript

// modules named like WIT-style packages ("extism:host") declare host function imports
declare module 'extism:host' { // namespace
  interface user {
    export myHostFunction1(p: I32, q: I32): I32;
    export myHostFunction2(p: I32): I64;
  } // which tees us up to implement "import { user } from 'extism:host'"
}

// The "main" module declares exports
declare module 'main' {
  // all extism exports have type () -> i32
  export function greet(): I32;
}
```
```